xls/public/c_api_dslx.cc

```diff
@@ -0,0 +1,290 @@
+  return cpp_module->GetTypeDefinitions().size();
+}
+
+xls_dslx_type_definition_kind xls_dslx_module_get_type_definition_kind(
+    struct xls_dslx_module* module, int64_t i) {
+  auto* cpp_module = reinterpret_cast<xls::dslx::Module*>(module);
+  xls::dslx::TypeDefinition cpp_type_definition =
+      cpp_module->GetTypeDefinitions().at(i);
+  return absl::visit(xls::Visitor{
+                         [](const xls::dslx::StructDef*) {
+                           return xls_dslx_type_definition_kind_struct_def;
+                         },
+                         [](const xls::dslx::EnumDef*) {
+                           return xls_dslx_type_definition_kind_enum_def;
+                         },
+                         [](const xls::dslx::TypeAlias*) {
+                           return xls_dslx_type_definition_kind_type_alias;
+                         },
+                         [](const xls::dslx::ColonRef*) {
+                           return xls_dslx_type_definition_kind_colon_ref;
+                         },
+                     },
+                     cpp_type_definition);
+}
+
+struct xls_dslx_struct_def* xls_dslx_module_get_type_definition_as_struct_def(
+    struct xls_dslx_module* module, int64_t i) {
+  auto* cpp_module = reinterpret_cast<xls::dslx::Module*>(module);
+  xls::dslx::TypeDefinition cpp_type_definition =
+      cpp_module->GetTypeDefinitions().at(i);
+  auto* cpp_struct_def = std::get<xls::dslx::StructDef*>(cpp_type_definition);
+  return reinterpret_cast<xls_dslx_struct_def*>(cpp_struct_def);
+}
+
+struct xls_dslx_enum_def* xls_dslx_module_get_type_definition_as_enum_def(
+    struct xls_dslx_module* module, int64_t i) {
+  auto* cpp_module = reinterpret_cast<xls::dslx::Module*>(module);
+  xls::dslx::TypeDefinition cpp_type_definition =
+      cpp_module->GetTypeDefinitions().at(i);
+  auto* cpp_enum_def = std::get<xls::dslx::EnumDef*>(cpp_type_definition);
+  return reinterpret_cast<xls_dslx_enum_def*>(cpp_enum_def);
+}
+
+char* xls_dslx_struct_def_get_identifier(struct xls_dslx_struct_def* n) {
+  auto* cpp_struct_def = reinterpret_cast<xls::dslx::StructDef*>(n);
+  std::string result = cpp_struct_def->identifier();
+  return xls::ToOwnedCString(result);
+}
+
+// -- enum_def
+
+char* xls_dslx_enum_def_get_identifier(struct xls_dslx_enum_def* n) {
+  auto* cpp_enum_def = reinterpret_cast<xls::dslx::EnumDef*>(n);
+  std::string result = cpp_enum_def->identifier();
+  return xls::ToOwnedCString(result);
+}
+
+int64_t xls_dslx_enum_def_get_member_count(struct xls_dslx_enum_def* n) {
+  auto* cpp_enum_def = reinterpret_cast<xls::dslx::EnumDef*>(n);
+  return static_cast<int64_t>(cpp_enum_def->values().size());
+}
+
+struct xls_dslx_enum_member* xls_dslx_enum_def_get_member(
+    struct xls_dslx_enum_def* n, int64_t i) {
+  auto* cpp_enum_def = reinterpret_cast<xls::dslx::EnumDef*>(n);
+  xls::dslx::EnumMember& cpp_member = cpp_enum_def->values().at(i);
```

Comment:
Do you actually need mutable access to this, or could you return a `const struct xls_dslx_enum_member*` ?

Reply:
I think it's OOK without const. We use const sparingly through XLS especially in the IR and DSLX AST. See the following for justification (on the IR side):

https://google.github.io/xls/xls_style/#:~:text=Unlike%20most%20data,non%2Dconst%20pointers.

const in C++ has it's uses but I think using it too agressively can be counter productive given it's viral nature,